README.md
[![Pub](https://img.shields.io/pub/v/ms_material_color.svg)](https://pub.dartlang.org/packages/ms_material_color)

# Overview

A Flutter lib to simplify creation of MaterialColor with shaders

## Whats MS_Material_Color do

Create a MsMaterialColor and gerenate 10 shader of base color

## Usage

Just import lib and use:

```dart
import 'package:ms_material_color/ms_material_color.dart';

//.. more flutter code here.
@override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Color Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(primarySwatch: MsMaterialColor(0xffffd700)),
      home: MyHomePage(title: 'MS MaterialColor Demo Home Page'),
    );
  }
//.. more flutter code here too, maybe not.
```

![Screenshot](./screenshot.png)


CHANGELOG.md
## [1.0.0] - 08/25/2021.

* Update to support nullsafety

## [0.1.0] - 01/13/2020.

* First Release.


lib/ms_material_color.dart
import 'package:flutter/material.dart';
import 'dart:math' show min;

/// `MsMaterialColor` is a simple solution to creation a `MaterialColor`
/// with 10 shades variants 50, 100, 200, ...900 use only one base color.
///
/// See also:
///
///  * [MaterialColor](https://api.flutter.dev/flutter/material/MaterialColor-class.html), which
///    defines a single color as well a color swatch with ten shades of the color..
class MsMaterialColor extends MaterialColor {
  MsMaterialColor(int primary)
      : super(primary, {
          050: _genColor(Color(primary), 0.05),
          100: _genColor(Color(primary), 0.2),
          200: _genColor(Color(primary), 0.4),
          300: _genColor(Color(primary), 0.6),
          400: _genColor(Color(primary), 0.8),
          500: _genColor(Color(primary), 1.0),
          600: _genColor(Color(primary), 1.2),
          700: _genColor(Color(primary), 1.4),
          800: _genColor(Color(primary), 1.6),
          900: _genColor(Color(primary), 1.8),
        });

  /// Construct a MsMaterialColor from a String with hex color
  ///
  /// Hex color example "`ffc0ba77`""
  ///
  /// `ff` is alpha value.
  /// `c0` is a red value.
  /// `ba` is a green value.
  /// `77` is a blue value.
  factory MsMaterialColor.fromHexString(String hexString) =>
      MsMaterialColor(int.parse(hexString, radix: 16));

  /// Construct a MsMaterialColor from the lower 8 bits of four integers.
  ///
  /// * `a` is the alpha value, with 0 being transparent and 255 being fully
  ///   opaque.
  /// * `r` is [red], from 0 to 255.
  /// * `g` is [green], from 0 to 255.
  /// * `b` is [blue], from 0 to 255.
  ///
  /// Out of range values are brought into range using modulo 255.
  factory MsMaterialColor.fromARGB(int a, int r, int g, int b) =>
      MsMaterialColor((((a & 0xff) << 24) |
              ((r & 0xff) << 16) |
              ((g & 0xff) << 8) |
              ((b & 0xff) << 0)) &
          0xFFFFFFFF);

  /// Create a MsMaterialColor from red, green, blue, and opacity, similar to `rgba()` in CSS.
  ///
  /// * `r` is [red], from 0 to 255.
  /// * `g` is [green], from 0 to 255.
  /// * `b` is [blue], from 0 to 255.
  /// * `opacity` is alpha channel of this color as a double, with 0.0 being
  ///   transparent and 1.0 being fully opaque.
  ///
  /// Out of range values are brought into range using modulo 255.
  factory MsMaterialColor.fromRGBO(int r, int g, int b, double opacity) =>
      MsMaterialColor(((((opacity * 0xff ~/ 1) & 0xff) << 24) |
              ((r & 0xff) << 16) |
              ((g & 0xff) << 8) |
              ((b & 0xff) << 0)) &
          0xFFFFFFFF);

  static Color _genColor(Color color, double shade) {
    var hue = HSVColor.fromColor(color);
    if (shade > 1.0) {
      return hue.withValue(min(1, (shade - 2).abs())).toColor();
    } else {
      return hue.withSaturation(shade).withValue(1).toColor();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:ms_material_color/ms_material_color.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Color Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(primarySwatch: MsMaterialColor(0xffffd700)),
      home: MyHomePage(title: 'MS MaterialColor Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: <Widget>[
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade50,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 50 - ${(Theme.of(context).primaryColor as MaterialColor).shade50}'),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade100,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 100 - ${(Theme.of(context).primaryColor as MaterialColor).shade100}'),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade200,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 200 - ${(Theme.of(context).primaryColor as MaterialColor).shade200}'),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade300,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 300 - ${(Theme.of(context).primaryColor as MaterialColor).shade300}'),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade400,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 400 - ${(Theme.of(context).primaryColor as MaterialColor).shade400}'),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade500,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 500 - ${(Theme.of(context).primaryColor as MaterialColor).shade500}'),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade600,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 600 - ${(Theme.of(context).primaryColor as MaterialColor).shade600}'),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade700,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 700 - ${(Theme.of(context).primaryColor as MaterialColor).shade700}'),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade800,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 800 - ${(Theme.of(context).primaryColor as MaterialColor).shade800}'),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              color: (Theme.of(context).primaryColor as MaterialColor).shade900,
              padding: EdgeInsets.all(8.0),
              child: Row(
                children: <Widget>[
                  Text(
                      'Shade 900 - ${(Theme.of(context).primaryColor as MaterialColor).shade900}'),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}
